### PR TITLE
Adding fwTPs to the readout, Fix for https://github.com/DUNE-DAQ/fdreadoutlibs/issues/9

### LINKED ISSUE #9: SW : Integrate FW trigger primitives into readout
This is a fast and slightly dirty way to integrate firmware TPs.
This corresponds to PR https://github.com/DUNE-DAQ/fdreadoutlibs/pull/10 

1. the ReadoutType will hold the data in a std::vector < std::uint8_t >.
2. the latency buffer and recording therefore just deals in FELIX frames. Note that a FELIX frame, with superchunking set to >1, will store multiple TP hit frames.
3. there is no conversion from the "old" TP hit frame format to the "new" TP hit frame format.
4. the post-processing step in RAWWIBTriggerPrimitiveProcessor will include extraction of TPs/hits from the FELIX frames, and stitching TPs together

### PR DESCRIPTION
**Pull request by hristovaivana**
This issue concerns only the TP FELIX link to readout interface. 
It corresponds to task https://github.com/DUNE-DAQ/fdreadoutlibs/issues/9

The interface for fwTPG and DS will be added in another PR. here https://github.com/DUNE-DAQ/fdreadoutlibs/pull/11

### CODE REVIEW

File: include/fdreadoutlibs/FDReadoutTypes.hpp [hunk 1]
```diff
@@ -502,37 +499,43 @@ struct RAW_WIB_TRIGGERPRIMITIVE_STRUCT
...
+
+    //unpack_timestamp(); // TODO 2022-04-15: slows down readout; timestamp not really needed here? 
+  }
+  int get_raw_tp_frame_chunksize()
+  {
+    return m_raw_tp_frame_chunksize;
   }
 
+private:
+void unpack_timestamp() {
```
Comment (jimbrooke) [inline include/fdreadoutlibs/FDReadoutTypes.hpp:521]: is this method still valid?  I am struggling to parse it.  I am guessing it just uses RawWIBTp to retrieve the _first_ (only) timestamp in the packet?  If so then fine.

File: include/fdreadoutlibs/wib/TPEmulatorModel.hpp [hunk 2]
```diff
@@ -286,6 +287,54 @@ class TPEmulatorModel : public readoutlibs::SourceEmulatorConcept
     }
     TLOG_DEBUG(TLVL_WORK_STEPS) << "Data generation thread " << m_this_link_number << " finished";
   }
```
Comment (jimbrooke) [inline include/fdreadoutlibs/wib/TPEmulatorModel.hpp]: this huge block of commented code seems like bad practise to me.  Also, is this the *same* code as RAWWIBTriggerPrimitiveProcessor::unpack_tpframe_version_2() ?  (If so, should this code be in a separate class, for re-use where required ?)

File: include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp [hunk 3]
```diff
@@ -298,6 +160,188 @@ void tp_unpack(frame_ptr fr)
...
+    // raw recording 
+    tp_dump_frame();
+
+    // stitch TP hits
+    tp_stitch();
+    offset += (2+nhits)*RAW_WIB_TP_SUBFRAME_SIZE;
+  }
+}
+
+void unpack_tpframe_version_2(frame_ptr fr)
```
Comment (jimbrooke) [inline include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp]: do we need this function ?  there is no production firmware running version 2 yet.

**Review by jimbrooke — COMMENTED** [review]
I put a few comments against line numbers in the diff.  Here are some more general ones.

- I have not reviewed code in this package previously.  But is there a reason ALL the code is in header files?  😱 
- RAW_WIB_TRIGGERPRIMITIVE_STRUCT contains a std::unique_ptr<dunedaq::detdataformats::wib::RawWIBTp> rwtp.  I thought we agreed in this version it would not use RawWIBTp at this stage, and instead just use a vector of char? (which I assume is std::vector<std::uint8_t> m_raw_tp_frame_chunk ).
- is RAW_WIB_TRIGGERPRIMITIVE_STRUCT::unpack_timestamp() actually required ?
- in RAWWIBTriggerPrimitiveProcessor it looks like tp_unpack() is bound to the pre-processor tasks.  I thought we said this would be done in post-processing ?
- are RAWWIBTriggerPrimitiveProcessor::tp_dump_frame() or RAWWIBTriggerPrimitiveProcessor::tp_stitch() actually used?

File: include/fdreadoutlibs/wib/TPEmulatorModel.hpp (hunk 2 above)
Comment (hristovaivana) [reply include/fdreadoutlibs/wib/TPEmulatorModel.hpp]: well spotted, thanks. I wanted to delete the commented out code but forgot before the PR was made.

File: include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp (hunk 3 above)
Comment (hristovaivana) [reply include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp]: good point! I assume that now we can take this decision here and discard any handling of the new TP fw format, and target the next real running setup which will use the old TP format. 
 
The new TP format was used for development only. 

Would it be a compromise to leave the new TP format code but completely remove it from the execution path? Or perhaps it would be better and cleaner to delete it?

File: include/fdreadoutlibs/FDReadoutTypes.hpp (hunk 1 above)
Comment (hristovaivana) [reply include/fdreadoutlibs/FDReadoutTypes.hpp:521]: To address the points in order:

- We can discuss this with Roland or as part of a more major code review, 
- The rwtp pointer is still there in order not to break code that still uses it, I left a comment about this, was it not clear?
- No, the timestamp is not required to record the TP frames. 
- No, I think tp_unpack is a separate post-processing queue which uses the RAW_WIB_TRIGGERPRIMITIVE_STRUCT-s from the pre-processing queue. 
- Yes, they are both used, one for (our backup) recording the TP frames and the other to stitch the hits in the raw TP frames.

Comment (hristovaivana) [reply include/fdreadoutlibs/FDReadoutTypes.hpp:521]: Yes, correct, unpack_timestamp() just extracts the very first timestamp in any FELIX packet.

File: include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp (hunk 3 above)
Comment (jimbrooke) [reply include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp]: I don't think this is a big problem, so let's leave it.

**Comment by jimbrooke** [comment]
Re : RAW_WIB_TRIGGERPRIMITIVE_STRUCT contains a RawWIBTp pointer.  @hristovaivana  I see the comment, but it is a TEST application that uses it.  So why do we not fix the test application?  (We're talking about flxlibs/test/apps/test_tp_elinkhandler_app.cxx right ?)

**Comment by jimbrooke** [comment]
Re : RAW_WIB_TRIGGERPRIMITIVE_STRUCT::unpack_timestamp() - @hristovaivana  you said two things which appear contradictory.  I think you are saying this function is used to unpack the timestamp, but that the timestamps are not used in our current applications.  Right?

What I REALLY don't understand is why get_first_timestamp() does not just return 16 bytes at a fixed offset from the start of the std::vector<uint8_t>.  It seems to me use of RawWIBTp and unpack_timestamp() are complicating this, but it could be super simple.  Right ?  (And again, I go back to why is RawWIBTp still in RAW_WIB_TRIGGERPRIMITIVE_STRUCT ?)

**Comment by jimbrooke** [comment]
Re : binding tp_unpack() to pre-processing tasks.  @hristovaivana I wrote in #9 my understanding of how to resolve the problem.  This PR seems to do something different from what I described.  Can you explain : 

1. what is wrong with #9, and/or why the unpacking should happen in the pre-processing?
2. how are the unpacked quantities accessed in post-processing?  Is it via the RawWIBTp pointer ?  Are we then duplicating the data before we put it into the latency buffer - or have I misunderstood ?

**Comment by jimbrooke** [comment]
Re : RAWWIBTriggerPrimitiveProcessor::tp_dump_frame()  and tp_stitch().  @hristovaivana I don't see how they can be used in this PR.  Are you saying they will be used in #11  ?

File: include/fdreadoutlibs/FDReadoutTypes.hpp [hunk 4]
```diff
@@ -431,67 +431,64 @@ struct RAW_WIB_TRIGGERPRIMITIVE_STRUCT
   RAW_WIB_TRIGGERPRIMITIVE_STRUCT()
   {
     m_raw_tp_frame_chunksize = 0;
+    m_first_timestamp = 0;
   }
 
-  using FrameType = dunedaq::detdataformats::wib::RawWIBTp;
+  using FrameType = RAW_WIB_TRIGGERPRIMITIVE_STRUCT;
```
Comment (roland-sipos) [inline include/fdreadoutlibs/FDReadoutTypes.hpp:437]: Why are we using the same aggregator type for FrameType in its aggregator struct/type?

File: include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp [hunk 5]
```diff
@@ -80,175 +80,37 @@ class RAWWIBTriggerPrimitiveProcessor
...
-    uint16_t padding = m_payload_wrapper.rwtp->get_padding_3(); // NOLINT
-
-    if (padding == 48879) { // padding hex is BEEF, new TP format
-      TLOG_DEBUG(TLVL_WORK_STEPS) << "Unpack raw WIB TP frame version 2.";
-      unpack_tpframe_version_2(fr, nhits, offset);
-
-    } else { // old TP format
-      TLOG_DEBUG(TLVL_WORK_STEPS) << "Unpack raw WIB TP frame version 1.";
-      unpack_tpframe_version_1(fr, offset);
+      fwrite(&rwtp->m_head, 1, sizeof(dunedaq::detdataformats::wib::TpHeader),
```
Comment (roland-sipos) [inline include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp]: I would appreciate to avoid using C fwrite. This is (mainly) C++ code, and we should only use native C code when necessary. Please move to fstream.

File: include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp [hunk 6]
```diff
@@ -80,175 +80,37 @@ class RAWWIBTriggerPrimitiveProcessor
...
     int nhits = rwtp->get_nhits();
     uint64_t ts_0 = rwtp->get_timestamp(); // NOLINT
-    rwtp->set_timestamp(ts_0); // NOLINT
+    //rwtp->set_timestamp(ts_0); // NOLINT
     uint8_t m_channel_no = rwtp->m_head.m_wire_no; // NOLINT
     uint8_t m_fiber_no = rwtp->m_head.m_fiber_no; // NOLINT
     for (int i = 0; i < nhits; ++i) {
-      TLOG_DEBUG(TLVL_WORK_STEPS) << "Processing raw TP hit " << i << " out of " << nhits << " hits in this frame.";
-      //TLOG() << "Processing raw TP hit " << i << " out of " << nhits << " hits in this frame.";
+      TLOG_DEBUG(TLVL_WORK_STEPS) << "Processing raw TP hit " << i << " out of " << nhits << " hits in this   frame.";
```
Comment (roland-sipos) [inline include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp]: Work_steps will be a way too high verbosity level for TP stitching! This will flood output.

**Comment by hristovaivana** [comment]
Hi Jim, 
**1. what is wrong with
https://github.com/DUNE-DAQ/fdreadoutlibs/issues/9, and/or why the unpacking should happen in the pre-processing?**

I removed tp_unpack() because it is not ready to send TPs in the form suitable for the trigger. But the place where tp_unpack() 
is currently located (as a placeholder for now) is what I would call a "post-processing" step. I might be misunderstanding it but 
this is my understanding.

**2. how are the unpacked quantities accessed in post-processing? Is it via the RawWIBTp pointer ? Are we then duplicating the data before we put it into the latency buffer - or have I misunderstood ?**

The unpacked quantities are accessed via 1) a pointer to the RAW_WIB_TRIGGERPRIMITIVE_STRUCT from the Latency Buffer from the pre-processing step and 2) extracting/storing  them into RawWIBTp pointer (this is the post-processing).
No, we are not duplicating the TP frames, once read in from the FELIX link they are stored in the Latency Buffer. This is what I call the pre-processing step, and it currently does nothing more than putting the FELIX data in a std::vector, and extracting the timestamp (only the first timestamp in a FELIX chunk). The Latency Buffer holds the RAW_WIB_TRIGGERPRIMITIVE_STRUCT.

The unpacking (conversion from  RAW_WIB_TRIGGERPRIMITIVE_STRUCT to RawWIBTp) is followed by the stitching, and this is the post-processing step, which takes the RAW_WIB_TRIGGERPRIMITIVE_STRUCT from the Latency Buffer. I initially put this in the PR because this point was included in your description of this PR. However, this functionality will be added in the next PR.

**Comment by hristovaivana** [comment]
**Re : RAWWIBTriggerPrimitiveProcessor::tp_dump_frame() and tp_stitch(). @hristovaivana I don't see how they can be used in this PR. Are you saying they will be used in https://github.com/DUNE-DAQ/fdreadoutlibs/pull/11 ?**

Yes, exactly.
The Latency Buffer is the separation line between the pre-processing and post-processing.

**Comment by jimbrooke** [comment]
@hristovaivana I am going to keep my replies in threads otherwise the conversation is hard to follow.

> Hi Jim, **1. what is wrong with #9, and/or why the unpacking should happen in the pre-processing?**
> 
> I removed tp_unpack() because it is not ready to send TPs in the form suitable for the trigger. But the place where tp_unpack() is currently located (as a placeholder for now) is what I would call a "post-processing" step. I might be misunderstanding it but this is my understanding.

Here's an example of a post-processing task (ie. SW TP finding) : 
https://github.com/DUNE-DAQ/fdreadoutlibs/blob/develop/include/fdreadoutlibs/wib/WIBFrameProcessor.hpp#L255

But in this PR you are using TaskRawDataProcessorModel::add_**preprocess**_task() : 
https://github.com/DUNE-DAQ/fdreadoutlibs/blob/hristova/firmware_tpg_integration/include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp#L54

> **2. how are the unpacked quantities accessed in post-processing? Is it via the RawWIBTp pointer ? Are we then duplicating the data before we put it into the latency buffer - or have I misunderstood ?**
> 
> The unpacked quantities are accessed via 1) a pointer to the RAW_WIB_TRIGGERPRIMITIVE_STRUCT from the Latency Buffer from the pre-processing step and 2) extracting/storing them into RawWIBTp pointer (this is the post-processing). No, we are not duplicating the TP frames, once read in from the FELIX link they are stored in the Latency Buffer. This is what I call the pre-processing step, and it currently does nothing more than putting the FELIX data in a std::vector, and extracting the timestamp (only the first timestamp in a FELIX chunk). The Latency Buffer holds the RAW_WIB_TRIGGERPRIMITIVE_STRUCT.
> 
> The unpacking (conversion from RAW_WIB_TRIGGERPRIMITIVE_STRUCT to RawWIBTp) is followed by the stitching, and this is the post-processing step, which takes the RAW_WIB_TRIGGERPRIMITIVE_STRUCT from the Latency Buffer. I initially put this in the PR because this point was included in your description of this PR. However, this functionality will be added in the next PR.

I didn't really follow this, but maybe we come back to that after we have understood point 1.

File: include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp (hunk 5 above)
Comment (jimbrooke) [reply include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp]: @hristovaivana I think this is resolved.  If so, can you resolve the conversation.

**Comment by hristovaivana** [comment]
Hi @jimbrooke ,

To my understanding the difference between add_preprocess_task and add_postprocess_task is that the former is suited for sequential processing of data from the links while the latter is parallelising the processing (via threads) when the task at hand allows for it.
I've chosen add_preprocess_task, initially, because I thought that the stithcing algorithm could not run as a parallel task, and I have not re-considered this decision afterwards.  

I would suggest to remove the reference to it completely from this tag, because it is not performing any functionality anyway. Then I will think about it again when addressing the 2nd pull request. 

Would you agree if I remove the following line from this PR:
https://github.com/DUNE-DAQ/fdreadoutlibs/blob/696dbe469f148cdf40e818e0aab5b4ee6e766abc/include/fdreadoutlibs/wib/RAWWIBTriggerPrimitiveProcessor.hpp#L54

**Comment by jimbrooke** [comment]
yes, let's do that. we can pick this up in PR #11.  my thoughts, for the record, are that unpacking/stitching/publishing in this version are all part of the same task, since the functionality goals are 1) recording and 2) publishing.  and i think stitching is very similar to SW TPG generation, so we can use that as a model (including parallelisation - which I guess can be done eg. by grouping channels to threads).

**Comment by hristovaivana** [comment]
Great, thanks Jim, I will remove the ambiguous registration of the processing task and also address Roland's comment indeed. I now understand his point and I will fix it. 
I will update this PR asap.

File: include/fdreadoutlibs/FDReadoutTypes.hpp (hunk 4 above)
Comment (hristovaivana) [reply include/fdreadoutlibs/FDReadoutTypes.hpp:437]: Hi @roland-sipos ,

Thanks for pointing this out. It is a good question indeed. I think it is a matter of choice and I made this 
choice now. It is currently similar to SW_WIB_TRIGGERPRIMITIVE_STRUCT.
I would suggest to keep this choice for now, however if you have a suggestion of other aggregator type, please let us know. The choice depends on how RAW_WIB_TRIGGERPRIMITVE_STRUCT is used, and in this PR there is not other critical functionality that relies on it. Please let us know what you think and how to proceed with this PR. 

I would suggest to address this comment in more detail in PR https://github.com/DUNE-DAQ/fdreadoutlibs/pull/11 unless you think it is a showstopper for https://github.com/DUNE-DAQ/fdreadoutlibs/pull/10 (this one).

**Review by jimbrooke — APPROVED** [review]
I am OK with these changes now.  But I propose we test with Shyam's changes to flxlibs before merging.
